### PR TITLE
openstack essential flavors and images support for VM creations -- part 1

### PR DESCRIPTION
**Pull request by yb01**
This is the first part of the VM creation to use the openstack flavors and images.

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+func GetFalvor(name string) (*FlavorType, error) {
+	if flavor, found := flavors[name]; found {
+		return &flavor, nil
+	}
+
+	return nil, fmt.Errorf("flavor %s, not found", name)
+}
+
+func ListFalvors() []*FlavorType {
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go:50]: This function returns array of pointers. No need to construct array every time.

File: staging/src/k8s.io/apiserver/pkg/util/openstack/images.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+func GetImage(name string) (*ImageType, error) {
+	if image, found := images[name]; found {
+		return &image, nil
+	}
+
+	return nil, fmt.Errorf("image %s, not found", name)
+}
+
+func ListImages() []*ImageType {
+	imageList := make([]*ImageType, len(flavors))
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/util/openstack/images.go:47]: Same here

File: staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go [hunk 3]
```diff
@@ -38,12 +38,21 @@ func init() {
 	}
 
 	POD_JSON_STRING_TEMPLATE = string(t)
+
+	initFlavorsCache()
+	initImagesCache()
+
+	fl := ListFalvors()
+	klog.(V6).Infof("built-in flavors: %v", fl)
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:46]: I don't know klog.(V6) work

Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go]: typo, fixed. thanks.

File: staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go (hunk 1 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go:50]: This is intentional to rebuild when the function is called. for couple of reasons:  flavors, images will be added in future; and also the map is used for the structure for quick GET() support hence the conventing to the slices from map values.

Comment (Sindica) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go:50]: In this code, all flavors are added at the init cache stage. There is no add function to support dynamic adding/delete flavors. I don't see why the list cannot be initialized and reused.

Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go:58]: fixed and add comment for the READ only support for now.

File: staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	Id          int
+	Name        string
+	MemoryMb    int
+	Vcpus       int
+	RootGb      int
+	EphemeralGb int
+}
+
+// for 130 release, only READ operation with static list of flavors
+func initFlavorsCache() {
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/util/openstack/flavors.go:35]: Question: I did not find the usage of initFlavorsCache and initImagesCache. When will they be called and initialize the cache?

**Review by Sindica — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/apiserver/pkg/util/openstack/images.go (hunk 2 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/images.go]: fixed as well.

**Pull request merged**